// vuestic-ui/dist/esm/src/components/va-navbar/VaNavbar.js
import { defineComponent, computed, openBlock, createElementBlock, normalizeStyle, createElementVNode, renderSlot, createCommentVNode } from "vue";
import "../../../vendor.js";
import { u as useColors } from "../../composables/useColors.js";
import "../../composables/useDropdown.js";
import { u as useTextColor } from "../../composables/useTextColor.js";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var VaNavbar_vue_vue_type_style_index_0_lang = "";
const _sfc_main = defineComponent({
  name: "VaNavbar",
  props: {
    color: { type: String, default: "secondary" },
    textColor: { type: String },
    shape: { type: Boolean, default: false }
  },
  setup(props) {
    const { getColor, shiftHSLAColor } = useColors();
    const { textColorComputed } = useTextColor();
    const color = computed(() => getColor(props.color));
    const shapeStyleComputed = computed(() => ({
      borderTopColor: shiftHSLAColor(color.value, { h: -1, s: -11, l: 10 })
    }));
    const computedStyle = computed(() => ({
      backgroundColor: color.value,
      color: textColorComputed.value,
      fill: textColorComputed.value
    }));
    return {
      computedStyle,
      shapeStyleComputed
    };
  }
});
const _hoisted_1 = { class: "va-navbar__content" };
const _hoisted_2 = {
  key: 0,
  class: "va-navbar__left"
};
const _hoisted_3 = {
  key: 1,
  class: "va-navbar__center"
};
const _hoisted_4 = {
  key: 2,
  class: "va-navbar__right"
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("header", {
    class: "va-navbar",
    style: normalizeStyle(_ctx.computedStyle)
  }, [
    createElementVNode("div", _hoisted_1, [
      _ctx.$slots.left ? (openBlock(), createElementBlock("div", _hoisted_2, [
        renderSlot(_ctx.$slots, "left")
      ])) : createCommentVNode("", true),
      _ctx.$slots.center ? (openBlock(), createElementBlock("div", _hoisted_3, [
        renderSlot(_ctx.$slots, "center")
      ])) : createCommentVNode("", true),
      _ctx.$slots.right ? (openBlock(), createElementBlock("div", _hoisted_4, [
        renderSlot(_ctx.$slots, "right")
      ])) : createCommentVNode("", true)
    ]),
    _ctx.shape ? (openBlock(), createElementBlock("div", {
      key: 0,
      class: "va-navbar__background-shape",
      style: normalizeStyle(_ctx.shapeStyleComputed)
    }, null, 4)) : createCommentVNode("", true)
  ], 4);
}
var _VaNavbar = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { _VaNavbar as _ };

import './VaNavbar.css';
