// vuestic-ui/dist/esm/src/components/va-card/VaCardActions.js
var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
import { defineComponent, openBlock, createElementBlock, normalizeClass, normalizeStyle, renderSlot } from "vue";
import { u as useAlignProps, a as useAlign } from "../../composables/useAlign.js";
import "../../../vendor.js";
import "colortranslator";
import "../../composables/useDropdown.js";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var VaCardActions_vue_vue_type_style_index_0_lang = "";
const _sfc_main = defineComponent({
  name: "VaCardActions",
  props: __spreadValues({}, useAlignProps),
  setup(props) {
    const { alignComputed } = useAlign(props);
    return {
      alignComputed
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: normalizeClass([_ctx.vertical ? "va-card__actions_vertical" : "va-card__actions"]),
    style: normalizeStyle(_ctx.alignComputed)
  }, [
    renderSlot(_ctx.$slots, "default")
  ], 6);
}
var VaCardActionsBase = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { VaCardActionsBase as V };

import './VaCardActions.css';
