// vuestic-ui/dist/esm/src/components/va-collapse/VaCollapse.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
import { defineComponent, shallowRef, ref, computed, resolveComponent, openBlock, createElementBlock, normalizeClass, createElementVNode, mergeProps, toHandlers, withKeys, renderSlot, normalizeProps, guardReactiveProps, createBlock, createCommentVNode, toDisplayString, createVNode, normalizeStyle } from "vue";
import { p as pick_1 } from "../../../vendor.js";
import { u as useBem } from "../../composables/useBem.js";
import { u as useColors } from "../../composables/useColors.js";
import { u as useKeyboardOnlyFocus } from "../../composables/useKeyboardOnlyFocus.js";
import "../../composables/useDropdown.js";
import { u as useResizeObserver } from "../../composables/useResizeObserver.js";
import { u as useSyncProp } from "../../composables/useSyncProp.js";
import { g as generateUniqueId } from "../../services/utils.js";
import { u as useTextColor } from "../../composables/useTextColor.js";
import { a as useAccordionItem } from "../va-accordion/hooks/useAccordion.js";
import { V as VaIcon } from "../va-icon/index.js";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var VaCollapse_vue_vue_type_style_index_0_lang = "";
const _sfc_main = defineComponent({
  name: "VaCollapse",
  components: {
    VaIcon
  },
  props: {
    modelValue: { type: Boolean, default: void 0 },
    disabled: { type: Boolean, default: false },
    header: { type: String, default: "" },
    icon: { type: String, default: "" },
    solid: { type: Boolean, default: false },
    color: { type: String, default: "background" },
    textColor: { type: String, default: "" },
    colorAll: { type: Boolean, default: false }
  },
  emits: ["focus", "update:modelValue"],
  setup(props, { emit, slots }) {
    const body = shallowRef();
    const [computedModelValue] = useSyncProp("modelValue", props, emit, false);
    const { getColor, getHoverColor } = useColors();
    const { accordionProps, toggle } = useAccordionItem(computedModelValue);
    const { textColorComputed } = useTextColor();
    const bodyHeight = ref();
    useResizeObserver([body], () => {
      var _a, _b;
      bodyHeight.value = (_b = (_a = body.value) == null ? void 0 : _a.clientHeight) != null ? _b : 0;
    });
    const height = computed(() => computedModelValue.value ? bodyHeight.value : 0);
    const { hasKeyboardFocus, keyboardFocusListeners } = useKeyboardOnlyFocus();
    const getTransition = () => {
      const duration = height.value / 1e3 * 0.2;
      return `${duration > 0.2 ? duration : 0.2}s`;
    };
    const getBackground = () => {
      return props.color && props.colorAll ? getHoverColor(getColor(props.color)) : "";
    };
    const uniqueId = computed(generateUniqueId);
    const headerIdComputed = computed(() => `header-${uniqueId.value}`);
    const panelIdComputed = computed(() => `panel-${uniqueId.value}`);
    const tabIndexComputed = computed(() => props.disabled ? -1 : 0);
    const headerAttributes = computed(() => ({
      id: headerIdComputed.value,
      tabindex: tabIndexComputed.value,
      "aria-controls": panelIdComputed.value,
      "aria-expanded": computedModelValue.value,
      "aria-disabled": props.disabled,
      role: "button"
    }));
    const computedClasses = useBem("va-collapse", () => __spreadProps(__spreadValues({}, pick_1(props, ["disabled", "solid"])), {
      expanded: computedModelValue.value,
      active: props.solid && computedModelValue.value,
      popout: !!(accordionProps.value.popout && computedModelValue.value),
      inset: !!(accordionProps.value.inset && computedModelValue.value)
    }));
    return {
      body,
      height,
      toggle,
      computedModelValue,
      hasKeyboardFocus,
      keyboardFocusListeners,
      textColorComputed,
      headerIdComputed,
      headerAttributes,
      panelIdComputed,
      tabIndexComputed,
      computedClasses,
      headerStyle: computed(() => ({
        paddingLeft: props.icon && 0,
        color: textColorComputed.value,
        backgroundColor: props.color ? getColor(props.color) : ""
      })),
      contentStyle: computed(() => {
        var _a;
        const hasContent = computedModelValue.value && !!((_a = slots.default) == null ? void 0 : _a.call(slots)[0]);
        return {
          visibility: hasContent ? "visible" : "hidden",
          height: `${height.value}px`,
          transitionDuration: getTransition(),
          background: hasContent ? getBackground() : ""
        };
      })
    };
  }
});
const _hoisted_1 = { class: "va-collapse__header__text" };
const _hoisted_2 = ["id", "aria-labelledby"];
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_va_icon = resolveComponent("va-icon");
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(["va-collapse", _ctx.computedClasses])
  }, [
    createElementVNode("div", mergeProps({ class: "va-collapse__header-wrapper" }, toHandlers(_ctx.keyboardFocusListeners), {
      onFocus: _cache[0] || (_cache[0] = ($event) => _ctx.$emit("focus")),
      onClick: _cache[1] || (_cache[1] = (...args) => _ctx.toggle && _ctx.toggle(...args)),
      onKeydown: [
        _cache[2] || (_cache[2] = withKeys((...args) => _ctx.toggle && _ctx.toggle(...args), ["enter"])),
        _cache[3] || (_cache[3] = withKeys((...args) => _ctx.toggle && _ctx.toggle(...args), ["space"]))
      ]
    }), [
      renderSlot(_ctx.$slots, "header", normalizeProps(guardReactiveProps({
        value: _ctx.computedModelValue,
        hasKeyboardFocus: _ctx.hasKeyboardFocus,
        bind: _ctx.headerAttributes,
        attributes: _ctx.headerAttributes
      })), () => [
        createElementVNode("div", mergeProps(_ctx.headerAttributes, {
          class: "va-collapse__header",
          style: _ctx.headerStyle
        }), [
          _ctx.icon ? (openBlock(), createBlock(_component_va_icon, {
            key: 0,
            class: "va-collapse__header__icon",
            name: _ctx.icon,
            color: _ctx.textColorComputed
          }, null, 8, ["name", "color"])) : createCommentVNode("", true),
          createElementVNode("div", _hoisted_1, toDisplayString(_ctx.header), 1),
          createVNode(_component_va_icon, {
            class: "va-collapse__header__icon",
            name: _ctx.computedModelValue ? "expand_less" : "expand_more",
            color: _ctx.textColorComputed
          }, null, 8, ["name", "color"])
        ], 16)
      ])
    ], 16),
    createElementVNode("div", {
      class: "va-collapse__body-wrapper",
      style: normalizeStyle(_ctx.contentStyle)
    }, [
      createElementVNode("div", {
        class: "va-collapse_body",
        ref: "body",
        role: "region",
        id: _ctx.panelIdComputed,
        "aria-labelledby": _ctx.headerIdComputed
      }, [
        renderSlot(_ctx.$slots, "default")
      ], 8, _hoisted_2)
    ], 4)
  ], 2);
}
var _VaCollapse = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { _VaCollapse as _ };

import './VaCollapse.css';
