// vuestic-ui/dist/esm/src/components/va-file-upload/VaFileUploadList/VaFileUploadList.js
import { defineComponent, computed, resolveComponent, openBlock, createBlock, normalizeClass, withCtx, createElementBlock, Fragment, renderList, createCommentVNode } from "vue";
import "../../va-accordion/index.js";
import "../../va-affix/index.js";
import "../../va-alert/index.js";
import "../../va-app-bar/index.js";
import "../../va-avatar/index.js";
import "../../va-backtop/index.js";
import "../../va-badge/index.js";
import "../../va-breadcrumbs/index.js";
import "../../va-button-dropdown/index.js";
import "../../va-button-group/index.js";
import "../../va-button-toggle/index.js";
import "../../va-button/index.js";
import "../../va-card/index.js";
import "../../va-carousel/index.js";
import "../../va-checkbox/index.js";
import "../../va-chip/index.js";
import "../../va-collapse/index.js";
import "../../va-color-indicator/index.js";
import "../../va-color-input/index.js";
import "../../va-color-palette/index.js";
import "../../va-config/VaConfig.js";
import "../../va-content/index.js";
import "../../va-counter/index.js";
import "../../va-data-table/index.js";
import "../../va-date-input/index.js";
import "../../va-date-picker/index.js";
import "../../va-divider/index.js";
import "../../va-dropdown/index.js";
import "../index.js";
import "../../va-form/index.js";
import "../../va-hover/index.js";
import "../../va-icon/index.js";
import "../../va-image/index.js";
import "../../va-infinite-scroll/index.js";
import "../../va-inner-loading/index.js";
import "../../va-input/index.js";
import { b as VaList } from "../../va-list/index.js";
import "../../va-modal/VaModal.js";
import "../../va-navbar/index.js";
import "../../va-option-list/index.js";
import "../../va-pagination/index.js";
import "../../va-parallax/index.js";
import "../../va-popover/index.js";
import "../../va-progress-bar/index.js";
import "../../va-progress-circle/index.js";
import "../../va-radio/index.js";
import "../../va-rating/index.js";
import "../../va-select/index.js";
import "../../va-separator/VaSeparator.js";
import "../../va-sidebar/index.js";
import "../../va-slider/index.js";
import "../../va-spacer/VaSpacer.js";
import "../../va-switch/index.js";
import "../../va-toast/index.js";
import "../../va-tabs/index.js";
import "../../va-time-input/index.js";
import "../../va-time-picker/index.js";
import "../../va-timeline/index.js";
import "../../va-tree-view/index.js";
import { V as VaFileUploadListItem } from "../VaFileUploadListItem/index.js";
import { V as VaFileUploadGalleryItem } from "../VaFileUploadGalleryItem/index.js";
import { V as VaFileUploadSingleItem } from "../VaFileUploadSingleItem/index.js";
import { _ as _export_sfc } from "../../../../plugin-vue_export-helper.js";
var VaFileUploadList_vue_vue_type_style_index_0_lang = "";
const _sfc_main = defineComponent({
  name: "VaFileUploadList",
  components: {
    VaList,
    VaFileUploadListItem,
    VaFileUploadGalleryItem,
    VaFileUploadSingleItem
  },
  emits: ["remove", "removeSingle"],
  props: {
    type: { type: String, default: "" },
    files: { type: Array, default: null },
    color: { type: String, default: "success" }
  },
  setup(props) {
    const filesList = computed(() => props.files.map(convertFile));
    const convertFile = (file) => ({
      name: file.name || file.url || "",
      size: formatSize(file.size),
      date: formatDate(new Date()),
      image: file
    });
    const formatSize = (bytes) => {
      if (bytes === 0) {
        return "0 Bytes";
      }
      if (!bytes) {
        return "";
      }
      const k = 1024;
      const sizes = ["Bytes", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"];
      const i = Math.floor(Math.log(bytes) / Math.log(k));
      return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + " " + sizes[i];
    };
    const formatDate = (date = new Date()) => {
      return date.toLocaleDateString("en-US", {
        hour: "2-digit",
        minute: "2-digit",
        month: "short",
        day: "numeric",
        year: "numeric"
      });
    };
    return {
      filesList
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_va_file_upload_list_item = resolveComponent("va-file-upload-list-item");
  const _component_va_file_upload_gallery_item = resolveComponent("va-file-upload-gallery-item");
  const _component_va_file_upload_single_item = resolveComponent("va-file-upload-single-item");
  const _component_va_list = resolveComponent("va-list");
  return openBlock(), createBlock(_component_va_list, {
    class: normalizeClass(["va-file-upload-list", `va-file-upload-list--${_ctx.type}`]),
    role: _ctx.type !== "single" ? "list" : void 0
  }, {
    default: withCtx(() => [
      _ctx.type === "list" ? (openBlock(true), createElementBlock(Fragment, { key: 0 }, renderList(_ctx.filesList, (file, index) => {
        return openBlock(), createBlock(_component_va_file_upload_list_item, {
          key: file.name,
          file,
          color: _ctx.color,
          role: "listitem",
          onRemove: ($event) => _ctx.$emit("remove", index)
        }, null, 8, ["file", "color", "onRemove"]);
      }), 128)) : createCommentVNode("", true),
      _ctx.type === "gallery" ? (openBlock(true), createElementBlock(Fragment, { key: 1 }, renderList(_ctx.filesList, (file, index) => {
        return openBlock(), createBlock(_component_va_file_upload_gallery_item, {
          key: file.name,
          file,
          color: _ctx.color,
          role: "listitem",
          onRemove: ($event) => _ctx.$emit("remove", index)
        }, null, 8, ["file", "color", "onRemove"]);
      }), 128)) : createCommentVNode("", true),
      _ctx.type === "single" && _ctx.filesList.length ? (openBlock(), createBlock(_component_va_file_upload_single_item, {
        key: 2,
        file: _ctx.filesList[_ctx.filesList.length - 1],
        onRemove: _cache[0] || (_cache[0] = ($event) => _ctx.$emit("removeSingle"))
      }, null, 8, ["file"])) : createCommentVNode("", true)
    ]),
    _: 1
  }, 8, ["role", "class"]);
}
var _VaFileUploadList = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { _VaFileUploadList as _ };

import './VaFileUploadList.css';
