// vuestic-ui/dist/esm/src/components/va-rating/components/VaRatingItemNumberButton.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
import { defineComponent, openBlock, createElementBlock, normalizeStyle, toDisplayString } from "vue";
import "../../../../vendor.js";
import "colortranslator";
import "../../../composables/useDropdown.js";
import { u as useSizeProps, a as useSize } from "../../../composables/useSize.js";
import { u as useVaRatingColorsProps, a as useVaRatingColors } from "../hooks/useVaRatingColors.js";
import { _ as _export_sfc } from "../../../../plugin-vue_export-helper.js";
const _sfc_main = defineComponent({
  name: "VaRatingItemNumberButton",
  props: __spreadProps(__spreadValues(__spreadValues({}, useVaRatingColorsProps), useSizeProps), {
    itemNumber: { type: Number, required: true },
    modelValue: { type: Number, required: true }
  }),
  setup(props) {
    return __spreadValues(__spreadValues({}, useVaRatingColors(props)), useSize(props, "VaRating"));
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("button", {
    class: "va-rating__number-item",
    tabindex: "-1",
    "aria-hidden": "true",
    style: normalizeStyle({
      background: _ctx.backgroundComputed,
      color: _ctx.textColorComputed,
      width: _ctx.sizeComputed,
      height: _ctx.sizeComputed,
      fontSize: _ctx.fontSizeComputed,
      borderRadius: `${parseInt(_ctx.fontSizeComputed) * 0.125}rem`
    })
  }, toDisplayString(_ctx.itemNumber), 5);
}
var VaRatingItemNumberButton = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { VaRatingItemNumberButton as V };
