// vuestic-ui/dist/esm/src/components/va-timeline/VaTimelineItem/VaTimelineItem.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
import { defineComponent, h } from "vue";
import { e as extractComponentProps, f as filterComponentProps } from "../../../utils/child-props.js";
import { V as VaTimelineSeparator } from "../VaTimelineSeparator/index.js";
var VaTimelineItem_vue_vue_type_style_index_0_lang = "";
const COMPONENT_NAME = "va-timeline-item";
const VaTimelineSeparatorProps = extractComponentProps(VaTimelineSeparator);
const _sfc_main = defineComponent({
  name: COMPONENT_NAME,
  props: __spreadProps(__spreadValues({}, VaTimelineSeparatorProps), {
    color: { type: String, default: "primary" },
    isFirst: { type: Boolean },
    isLast: { type: Boolean },
    inverted: { type: Boolean }
  }),
  setup(props, { slots }) {
    const children = [
      h(VaTimelineSeparator, __spreadValues({}, filterComponentProps(props, VaTimelineSeparatorProps).value))
    ];
    const before = props.inverted ? slots.after : slots.before;
    if (before) {
      children.unshift(h("div", { class: `${COMPONENT_NAME}__before` }, before()));
    }
    const after = props.inverted ? slots.before : slots.after;
    if (after) {
      children.push(h("div", { class: `${COMPONENT_NAME}__after` }, after()));
    }
    return () => h("div", {
      class: [
        { [COMPONENT_NAME]: true },
        { [`${COMPONENT_NAME}--vertical`]: props.vertical },
        { [`${COMPONENT_NAME}--is-first`]: props.isFirst },
        { [`${COMPONENT_NAME}--is-last`]: props.isLast }
      ]
    }, children);
  }
});
export { _sfc_main as _ };

import './VaTimelineItem.css';
