// vuestic-ui/dist/esm/src/components/va-timeline/VaTimelineSeparator/VaTimelineSeparator.js
import { defineComponent, h } from "vue";
import "../../../../vendor.js";
import { u as useColors } from "../../../composables/useColors.js";
import "../../../composables/useDropdown.js";
var VaTimelineSeparator_vue_vue_type_style_index_0_lang = "";
const componentName = "va-timeline-separator";
const _sfc_main = defineComponent({
  name: componentName,
  props: {
    color: { type: String, default: "primary" },
    vertical: { type: Boolean },
    active: { type: Boolean },
    activePrevious: { type: Boolean },
    activeNext: { type: Boolean }
  },
  setup(props) {
    const { getColor } = useColors();
    return () => h("div", {
      class: {
        [componentName]: true,
        [`${componentName}--vertical`]: props.vertical
      }
    }, [
      h("div", {
        class: {
          [`${componentName}__line`]: true,
          [`${componentName}__line--active`]: props.activePrevious
        },
        style: {
          backgroundColor: getColor(props.activePrevious ? props.color : "divider")
        }
      }),
      h("div", {
        class: {
          [`${componentName}__center`]: true,
          [`${componentName}__center--active`]: props.active
        },
        style: {
          backgroundColor: getColor(props.active ? props.color : "divider")
        }
      }),
      h("div", {
        class: {
          [`${componentName}__line`]: true,
          [`${componentName}__line--active`]: props.activeNext
        },
        style: {
          backgroundColor: getColor(props.activeNext ? props.color : "divider")
        }
      })
    ]);
  }
});
export { _sfc_main as _ };

import './VaTimelineSeparator.css';
