// vuestic-ui/dist/esm/src/components/va-list/VaListItemSection.js
import { defineComponent, computed, openBlock, createElementBlock, normalizeClass, renderSlot } from "vue";
import { _ as _export_sfc } from "../../../plugin-vue_export-helper.js";
var VaListItemSection_vue_vue_type_style_index_0_lang = "";
const _sfc_main = defineComponent({
  name: "VaListItemSection",
  props: {
    icon: { type: Boolean, default: false },
    avatar: { type: Boolean, default: false }
  },
  setup: (props) => ({
    computedClass: computed(() => ({
      "va-list-item-section--main": !props.icon && !props.avatar,
      "va-list-item-section--icon": props.icon,
      "va-list-item-section--avatar": props.avatar
    }))
  })
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(["va-list-item-section", _ctx.computedClass])
  }, [
    renderSlot(_ctx.$slots, "default")
  ], 2);
}
var _VaListItemSection = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { _VaListItemSection as _ };

import './VaListItemSection.css';
