// vuestic-ui/dist/esm/src/components/va-input/components/VaMessageListWrapper.js
var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
import { defineComponent, computed, resolveComponent, openBlock, createElementBlock, renderSlot, createVNode } from "vue";
import "../../../../vendor.js";
import "colortranslator";
import "../../../composables/useDropdown.js";
import { u as useValidationProps } from "../../../composables/useValidation.js";
import { V as VaMessageList } from "./VaMessageList/index.js";
import { _ as _export_sfc } from "../../../../plugin-vue_export-helper.js";
var VaMessageListWrapper_vue_vue_type_style_index_0_lang = "";
const _sfc_main = defineComponent({
  name: "VaMessageListWrapper",
  components: { VaMessageList },
  props: __spreadValues({}, useValidationProps),
  setup(props) {
    return {
      messagesColor: computed(() => {
        if (props.error) {
          return "danger";
        }
        if (props.success) {
          return "success";
        }
        return "";
      }),
      messagesComputed: computed(() => props.error ? props.errorMessages : props.messages),
      errorLimit: computed(() => props.error ? Number(props.errorCount) : 99)
    };
  }
});
const _hoisted_1 = { class: "va-message-list-wrapper" };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_va_message_list = resolveComponent("va-message-list");
  return openBlock(), createElementBlock("div", _hoisted_1, [
    renderSlot(_ctx.$slots, "default"),
    createVNode(_component_va_message_list, {
      color: _ctx.messagesColor,
      limit: _ctx.errorLimit,
      "model-value": _ctx.messagesComputed
    }, null, 8, ["color", "limit", "model-value"])
  ]);
}
var VaMessageListWrapper = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { VaMessageListWrapper as V };

import './VaMessageListWrapper.css';
