// vuestic-ui/dist/esm/src/components/va-file-upload/VaFileUploadSingleItem/VaFileUploadSingleItem.js
import { defineComponent, resolveComponent, openBlock, createBlock, withCtx, createVNode, createElementVNode, toDisplayString, createCommentVNode, createTextVNode } from "vue";
import "../../../../vendor.js";
import "colortranslator";
import "../../../composables/useDropdown.js";
import { u as useStrictInject } from "../../../composables/useStrictInject.js";
import "../../va-accordion/index.js";
import "../../va-affix/index.js";
import "../../va-alert/index.js";
import "../../va-app-bar/index.js";
import "../../va-avatar/index.js";
import "../../va-backtop/index.js";
import "../../va-badge/index.js";
import "../../va-breadcrumbs/index.js";
import "../../va-button-dropdown/index.js";
import "../../va-button-group/index.js";
import "../../va-button-toggle/index.js";
import { V as VaButton } from "../../va-button/index.js";
import "../../va-card/index.js";
import "../../va-carousel/index.js";
import "../../va-checkbox/index.js";
import "../../va-chip/index.js";
import "../../va-collapse/index.js";
import "../../va-color-indicator/index.js";
import "../../va-color-input/index.js";
import "../../va-color-palette/index.js";
import "../../va-config/VaConfig.js";
import "../../va-content/index.js";
import "../../va-counter/index.js";
import "../../va-data-table/index.js";
import "../../va-date-input/index.js";
import "../../va-date-picker/index.js";
import "../../va-divider/index.js";
import "../../va-dropdown/index.js";
import "../index.js";
import "../../va-form/index.js";
import "../../va-hover/index.js";
import "../../va-icon/index.js";
import "../../va-image/index.js";
import "../../va-infinite-scroll/index.js";
import "../../va-inner-loading/index.js";
import "../../va-input/index.js";
import { V as VaListItem, a as VaListItemSection } from "../../va-list/index.js";
import "../../va-modal/VaModal.js";
import "../../va-navbar/index.js";
import "../../va-option-list/index.js";
import "../../va-pagination/index.js";
import "../../va-parallax/index.js";
import "../../va-popover/index.js";
import "../../va-progress-bar/index.js";
import "../../va-progress-circle/index.js";
import "../../va-radio/index.js";
import "../../va-rating/index.js";
import "../../va-select/index.js";
import "../../va-separator/VaSeparator.js";
import "../../va-sidebar/index.js";
import "../../va-slider/index.js";
import "../../va-spacer/VaSpacer.js";
import "../../va-switch/index.js";
import "../../va-toast/index.js";
import "../../va-tabs/index.js";
import "../../va-time-input/index.js";
import "../../va-time-picker/index.js";
import "../../va-timeline/index.js";
import "../../va-tree-view/index.js";
import { V as VaFileUploadKey } from "../types.js";
import { _ as _export_sfc } from "../../../../plugin-vue_export-helper.js";
var VaFileUploadSingleItem_vue_vue_type_style_index_0_lang = "";
const INJECTION_ERROR_MESSAGE = "The VaFileUploadSingleItem component should be used in the context of VaFileUpload component";
const _sfc_main = defineComponent({
  name: "VaFileUploadSingleItem",
  components: {
    VaButton,
    VaListItem,
    VaListItemSection
  },
  emits: ["remove"],
  props: {
    file: { type: Object, default: null }
  },
  setup: () => ({
    disabled: useStrictInject(VaFileUploadKey, INJECTION_ERROR_MESSAGE).disabled
  })
});
const _hoisted_1 = { class: "va-file-upload-single-item__name" };
const _hoisted_2 = /* @__PURE__ */ createTextVNode(" Delete ");
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_va_button = resolveComponent("va-button");
  const _component_va_list_item_section = resolveComponent("va-list-item-section");
  const _component_va_list_item = resolveComponent("va-list-item");
  return openBlock(), createBlock(_component_va_list_item, {
    disabled: _ctx.disabled,
    "aria-disabled": _ctx.disabled,
    class: "va-file-upload-single-item",
    tabindex: "-1"
  }, {
    default: withCtx(() => [
      createVNode(_component_va_list_item_section, { class: "va-file-upload-single-item__content" }, {
        default: withCtx(() => [
          createElementVNode("div", _hoisted_1, toDisplayString(_ctx.file && _ctx.file.name), 1),
          !_ctx.disabled ? (openBlock(), createBlock(_component_va_button, {
            key: 0,
            class: "va-file-upload-single-item__button",
            "aria-label": "remove file",
            size: "small",
            color: "danger",
            flat: "",
            onClick: _cache[0] || (_cache[0] = ($event) => _ctx.$emit("remove"))
          }, {
            default: withCtx(() => [
              _hoisted_2
            ]),
            _: 1
          })) : createCommentVNode("", true)
        ]),
        _: 1
      })
    ]),
    _: 1
  }, 8, ["disabled", "aria-disabled"]);
}
var _VaFileUploadSingleItem = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { _VaFileUploadSingleItem as _ };

import './VaFileUploadSingleItem.css';
