// vuestic-ui/dist/esm/src/components/va-file-upload/VaFileUploadGalleryItem/VaFileUploadGalleryItem.js
import { defineComponent, ref, computed, onMounted, watch, resolveComponent, openBlock, createBlock, normalizeClass, withCtx, createVNode, createElementBlock, createCommentVNode, createElementVNode, normalizeStyle, toDisplayString } from "vue";
import { d as colorToRgba } from "../../../services/color-config/color-functions.js";
import { u as useBem } from "../../../composables/useBem.js";
import { a as useFocus } from "../../../composables/useFocus.js";
import "../../../../vendor.js";
import "../../../composables/useDropdown.js";
import { u as useStrictInject } from "../../../composables/useStrictInject.js";
import { V as VaFileUploadKey } from "../types.js";
import "../../va-accordion/index.js";
import "../../va-affix/index.js";
import "../../va-alert/index.js";
import "../../va-app-bar/index.js";
import "../../va-avatar/index.js";
import "../../va-backtop/index.js";
import "../../va-badge/index.js";
import "../../va-breadcrumbs/index.js";
import "../../va-button-dropdown/index.js";
import "../../va-button-group/index.js";
import "../../va-button-toggle/index.js";
import { V as VaButton } from "../../va-button/index.js";
import "../../va-card/index.js";
import "../../va-carousel/index.js";
import "../../va-checkbox/index.js";
import "../../va-chip/index.js";
import "../../va-collapse/index.js";
import "../../va-color-indicator/index.js";
import "../../va-color-input/index.js";
import "../../va-color-palette/index.js";
import "../../va-config/VaConfig.js";
import "../../va-content/index.js";
import "../../va-counter/index.js";
import "../../va-data-table/index.js";
import "../../va-date-input/index.js";
import "../../va-date-picker/index.js";
import "../../va-divider/index.js";
import "../../va-dropdown/index.js";
import "../index.js";
import "../../va-form/index.js";
import "../../va-hover/index.js";
import "../../va-icon/index.js";
import "../../va-image/index.js";
import "../../va-infinite-scroll/index.js";
import "../../va-inner-loading/index.js";
import "../../va-input/index.js";
import { V as VaListItem, a as VaListItemSection } from "../../va-list/index.js";
import "../../va-modal/VaModal.js";
import "../../va-navbar/index.js";
import "../../va-option-list/index.js";
import "../../va-pagination/index.js";
import "../../va-parallax/index.js";
import "../../va-popover/index.js";
import "../../va-progress-bar/index.js";
import "../../va-progress-circle/index.js";
import "../../va-radio/index.js";
import "../../va-rating/index.js";
import "../../va-select/index.js";
import "../../va-separator/VaSeparator.js";
import "../../va-sidebar/index.js";
import "../../va-slider/index.js";
import "../../va-spacer/VaSpacer.js";
import "../../va-switch/index.js";
import "../../va-toast/index.js";
import "../../va-tabs/index.js";
import "../../va-time-input/index.js";
import "../../va-time-picker/index.js";
import "../../va-timeline/index.js";
import "../../va-tree-view/index.js";
import { V as VaFileUploadUndo } from "../VaFileUploadUndo/index.js";
import { _ as _export_sfc } from "../../../../plugin-vue_export-helper.js";
var VaFileUploadGalleryItem_vue_vue_type_style_index_0_lang = "";
const INJECTION_ERROR_MESSAGE = "The VaFileUploadGalleryItem component should be used in the context of VaFileUpload component";
const _sfc_main = defineComponent({
  name: "VaFileUploadGalleryItem",
  components: {
    VaFileUploadUndo,
    VaButton,
    VaListItem,
    VaListItemSection
  },
  emits: ["remove"],
  props: {
    file: { type: Object, default: null },
    color: { type: String, default: "success" }
  },
  setup(props, { emit }) {
    const {
      undo,
      disabled,
      undoDuration
    } = useStrictInject(VaFileUploadKey, INJECTION_ERROR_MESSAGE);
    const { isFocused, onFocus, onBlur } = useFocus();
    const previewImage = ref("");
    const removed = ref(false);
    const overlayStylesComputed = computed(() => ({
      backgroundColor: colorToRgba(props.color, 0.7)
    }));
    const classesComputed = useBem("va-file-upload-gallery-item", () => ({
      notImage: !previewImage.value,
      focused: isFocused.value,
      undo: removed.value
    }));
    const removeImage = () => {
      var _a;
      if (undo.value) {
        removed.value = true;
        setTimeout(() => {
          if (!removed.value) {
            return;
          }
          emit("remove");
          removed.value = false;
        }, (_a = undoDuration.value) != null ? _a : 0);
      } else {
        emit("remove");
        removed.value = false;
      }
    };
    const recoverImage = () => {
      removed.value = false;
    };
    const convertToImg = () => {
      if (!props.file.name || !props.file.image) {
        return;
      }
      if (props.file.image.url) {
        previewImage.value = props.file.image.url;
      } else if (props.file.image instanceof File) {
        const reader = new FileReader();
        reader.readAsDataURL(props.file.image);
        reader.onload = (e) => {
          var _a, _b;
          if (((_a = e.target) == null ? void 0 : _a.result).includes("image")) {
            previewImage.value = (_b = e.target) == null ? void 0 : _b.result;
          }
        };
      }
    };
    onMounted(convertToImg);
    watch(() => props.file, convertToImg);
    return {
      undo,
      removed,
      disabled,
      isFocused,
      previewImage,
      classesComputed,
      overlayStylesComputed,
      onBlur,
      onFocus,
      removeImage,
      recoverImage
    };
  }
});
const _hoisted_1 = ["src", "alt"];
const _hoisted_2 = { class: "va-file-upload-gallery-item__overlay" };
const _hoisted_3 = ["title"];
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_va_file_upload_undo = resolveComponent("va-file-upload-undo");
  const _component_va_list_item_section = resolveComponent("va-list-item-section");
  const _component_va_button = resolveComponent("va-button");
  const _component_va_list_item = resolveComponent("va-list-item");
  return openBlock(), createBlock(_component_va_list_item, {
    class: normalizeClass(["va-file-upload-gallery-item", _ctx.classesComputed]),
    tabindex: "-1",
    disabled: _ctx.disabled,
    "aria-disabled": _ctx.disabled,
    onFocus: _ctx.onFocus,
    onBlur: _ctx.onBlur
  }, {
    default: withCtx(() => [
      _ctx.removed && _ctx.undo ? (openBlock(), createBlock(_component_va_list_item_section, { key: 0 }, {
        default: withCtx(() => [
          createVNode(_component_va_file_upload_undo, {
            vertical: "",
            onRecover: _ctx.recoverImage
          }, null, 8, ["onRecover"])
        ]),
        _: 1
      })) : (openBlock(), createBlock(_component_va_list_item_section, { key: 1 }, {
        default: withCtx(() => [
          _ctx.previewImage ? (openBlock(), createElementBlock("img", {
            key: 0,
            src: _ctx.previewImage,
            alt: _ctx.file.name || "",
            class: "va-file-upload-gallery-item__image"
          }, null, 8, _hoisted_1)) : createCommentVNode("", true),
          createElementVNode("div", _hoisted_2, [
            createElementVNode("div", {
              class: "va-file-upload-gallery-item__overlay-background",
              style: normalizeStyle(_ctx.overlayStylesComputed)
            }, null, 4),
            _ctx.file && _ctx.file.name ? (openBlock(), createElementBlock("div", {
              key: 0,
              class: "va-file-upload-gallery-item__name",
              title: _ctx.file.name
            }, toDisplayString(_ctx.file.name), 9, _hoisted_3)) : createCommentVNode("", true),
            !_ctx.disabled ? (openBlock(), createBlock(_component_va_button, {
              key: 1,
              flat: "",
              color: "danger",
              icon: "delete_outline",
              class: "va-file-upload-gallery-item__delete",
              "aria-label": "remove file",
              onClick: _ctx.removeImage,
              onFocus: _ctx.onFocus,
              onBlur: _ctx.onBlur
            }, null, 8, ["onClick", "onFocus", "onBlur"])) : createCommentVNode("", true)
          ])
        ]),
        _: 1
      }))
    ]),
    _: 1
  }, 8, ["class", "disabled", "aria-disabled", "onFocus", "onBlur"]);
}
var _VaFileUploadGalleryItem = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { _VaFileUploadGalleryItem as _ };

import './VaFileUploadGalleryItem.css';
