// @amcharts/amcharts5/.internal/charts/stock/indicators/BollingerBands.js
import { __extends } from "tslib";
import { MovingAverage } from "./MovingAverage";
import { LineSeries } from "../../xy/series/LineSeries";
import * as $array from "../../../core/util/Array";
/**
 * An implementation of a [[StockChart]] indicator.
 *
 * @see {@link https://www.amcharts.com/docs/v5/charts/stock/indicators/} for more info
 */
var BollingerBands = /** @class */ (function (_super) {
    __extends(BollingerBands, _super);
    function BollingerBands() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Indicator series for the upper band.
         */
        Object.defineProperty(_this, "upperBandSeries", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * Indicator series for the lower band.
         */
        Object.defineProperty(_this, "lowerBandSeries", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(_this, "_editableSettings", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: [{
                    key: "period",
                    name: _this.root.language.translateAny("Period"),
                    type: "number"
                }, {
                    key: "upperColor",
                    name: _this.root.language.translateAny("Upper"),
                    type: "color"
                }, {
                    key: "seriesColor",
                    name: _this.root.language.translateAny("Average"),
                    type: "color"
                }, {
                    key: "lowerColor",
                    name: _this.root.language.translateAny("Lower"),
                    type: "color"
                }, {
                    key: "field",
                    name: _this.root.language.translateAny("Field"),
                    type: "dropdown",
                    options: ["open", "close", "low", "high", "hl/2", "hlc/3", "hlcc/4", "ohlc/4"]
                }, {
                    key: "type",
                    name: _this.root.language.translateAny("Type"),
                    type: "dropdown",
                    options: ["simple", "weighted", "exponential", "dema", "tema"]
                }]
        });
        return _this;
    }
    Object.defineProperty(BollingerBands.prototype, "_afterNew", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var stockSeries = this.get("stockSeries");
            var chart = stockSeries.chart;
            if (chart) {
                var upperBandSeries = chart.series.push(LineSeries.new(this._root, {
                    valueXField: "valueX",
                    valueYField: "upper",
                    openValueYField: "lower",
                    calculateAggregates: true,
                    xAxis: stockSeries.get("xAxis"),
                    yAxis: stockSeries.get("yAxis"),
                    groupDataDisabled: true,
                    themeTags: ["bollingerbands", "upper"]
                }));
                upperBandSeries.fills.template.set("visible", true);
                upperBandSeries.setPrivate("baseValueSeries", stockSeries);
                this.upperBandSeries = upperBandSeries;
                var lowerBandSeries = chart.series.push(LineSeries.new(this._root, {
                    valueXField: "valueX",
                    valueYField: "lower",
                    calculateAggregates: true,
                    xAxis: stockSeries.get("xAxis"),
                    yAxis: stockSeries.get("yAxis"),
                    groupDataDisabled: true,
                    themeTags: ["bollingerbands", "lower"]
                }));
                lowerBandSeries.setPrivate("baseValueSeries", stockSeries);
                this.lowerBandSeries = lowerBandSeries;
            }
            _super.prototype._afterNew.call(this);
            this.series.addTag("bollingerbands");
            this.series._applyThemes();
        }
    });
    Object.defineProperty(BollingerBands.prototype, "_updateChildren", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            _super.prototype._updateChildren.call(this);
            if (this.isDirty("upperColor")) {
                var color = this.get("upperColor");
                var upperBandSeries = this.upperBandSeries;
                upperBandSeries.set("stroke", color);
                upperBandSeries.set("fill", color);
                upperBandSeries.strokes.template.set("stroke", color);
                this._updateSeriesColor(upperBandSeries, color, "upperColor");
            }
            if (this.isDirty("lowerColor")) {
                var color = this.get("lowerColor");
                var lowerBandSeries = this.lowerBandSeries;
                lowerBandSeries.set("stroke", color);
                lowerBandSeries.strokes.template.set("stroke", color);
                this._updateSeriesColor(lowerBandSeries, color, "lowerColor");
            }
            this.setCustomData("standardDeviations", this.get("standardDeviations"));
        }
    });
    /**
     * @ignore
     */
    Object.defineProperty(BollingerBands.prototype, "prepareData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _this = this;
            _super.prototype.prepareData.call(this);
            if (this.series) {
                var period_1 = this.get("period", 20);
                var stockSeries = this.get("stockSeries");
                var dataItems_1 = stockSeries.dataItems;
                var standardDeviations_1 = this.get("standardDeviations", 2);
                var smaData = this.series.data.values;
                var i_1 = 0;
                $array.each(smaData, function (dataItem) {
                    if (i_1 >= period_1 - 1) {
                        var mean = dataItem.ma;
                        var stdSum = 0;
                        for (var j = i_1 - period_1 + 1; j <= i_1; j++) {
                            var di = dataItems_1[j];
                            var diValue = _this._getValue(di);
                            if (diValue != null) {
                                stdSum += Math.pow(diValue - mean, 2);
                            }
                        }
                        var std = Math.sqrt(stdSum / period_1);
                        var lower = mean - standardDeviations_1 * std;
                        var upper = mean + standardDeviations_1 * std;
                        dataItem.upper = upper;
                        dataItem.lower = lower;
                    }
                    i_1++;
                });
                this.upperBandSeries.data.setAll(smaData);
                this.lowerBandSeries.data.setAll(smaData);
            }
        }
    });
    Object.defineProperty(BollingerBands.prototype, "_dispose", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            this.upperBandSeries.dispose();
            this.lowerBandSeries.dispose();
            _super.prototype._dispose.call(this);
        }
    });
    Object.defineProperty(BollingerBands, "className", {
        enumerable: true,
        configurable: true,
        writable: true,
        value: "BollingerBands"
    });
    Object.defineProperty(BollingerBands, "classNames", {
        enumerable: true,
        configurable: true,
        writable: true,
        value: MovingAverage.classNames.concat([BollingerBands.className])
    });
    return BollingerBands;
}(MovingAverage));
export { BollingerBands };
