// @amcharts/amcharts5/.internal/charts/stock/indicators/ChaikinMoneyFlow.js
import { __extends } from "tslib";
import { ChartIndicator } from "./ChartIndicator";
import { LineSeries } from "../../xy/series/LineSeries";
import * as $array from "../../../core/util/Array";
/**
 * An implementation of a [[StockChart]] indicator.
 *
 * @see {@link https://www.amcharts.com/docs/v5/charts/stock/indicators/} for more info
 */
var ChaikinMoneyFlow = /** @class */ (function (_super) {
    __extends(ChaikinMoneyFlow, _super);
    function ChaikinMoneyFlow() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        Object.defineProperty(_this, "_editableSettings", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: [{
                    key: "period",
                    name: _this.root.language.translateAny("Period"),
                    type: "number"
                }, {
                    key: "seriesColor",
                    name: _this.root.language.translateAny("Color"),
                    type: "color"
                }]
        });
        Object.defineProperty(_this, "_themeTag", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "chaikinmoneyflow"
        });
        return _this;
    }
    Object.defineProperty(ChaikinMoneyFlow.prototype, "_createSeries", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            return this.panel.series.push(LineSeries.new(this._root, {
                xAxis: this.xAxis,
                yAxis: this.yAxis,
                valueXField: "valueX",
                valueYField: "cmf",
                stroke: this.get("seriesColor"),
                fill: undefined
            }));
        }
    });
    Object.defineProperty(ChaikinMoneyFlow.prototype, "_prepareChildren", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            if (this.isDirty("volumeSeries")) {
                this._dataDirty = true;
            }
            _super.prototype._prepareChildren.call(this);
        }
    });
    /**
     * @ignore
     */
    Object.defineProperty(ChaikinMoneyFlow.prototype, "prepareData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            if (this.series) {
                var dataItems = this.get("stockSeries").dataItems;
                var volumeSeries_1 = this.get("volumeSeries");
                this.setRaw("field", "close");
                var data_1 = this._getDataArray(dataItems);
                var i_1 = 0;
                $array.each(dataItems, function (dataItem) {
                    var close = dataItem.get("valueY");
                    if (close != null) {
                        var low = dataItem.get("lowValueY", close);
                        var high = dataItem.get("highValueY", close);
                        var volume = 1;
                        var volumeDI = volumeSeries_1.dataItems[i_1];
                        if (volumeDI) {
                            volume = volumeDI.get("valueY", 1);
                        }
                        var mf = ((close - low) - (high - close)) / (high - low);
                        if (high == low) {
                            mf = 0;
                        }
                        var mfv = mf * volume;
                        data_1[i_1].mfv = mfv;
                        data_1[i_1].volume = volume;
                    }
                    i_1++;
                });
                var period_1 = this.get("period", 20);
                i_1 = 0;
                $array.each(data_1, function (dataItem) {
                    if (i_1 >= period_1 - 1) {
                        var mfv = 0;
                        var volume = 0;
                        for (var j = i_1; j > i_1 - period_1; j--) {
                            mfv += data_1[j].mfv;
                            volume += data_1[j].volume;
                        }
                        dataItem.cmf = mfv / volume;
                    }
                    i_1++;
                });
                this.series.data.setAll(data_1);
            }
        }
    });
    Object.defineProperty(ChaikinMoneyFlow, "className", {
        enumerable: true,
        configurable: true,
        writable: true,
        value: "ChaikinMoneyFlow"
    });
    Object.defineProperty(ChaikinMoneyFlow, "classNames", {
        enumerable: true,
        configurable: true,
        writable: true,
        value: ChartIndicator.classNames.concat([ChaikinMoneyFlow.className])
    });
    return ChaikinMoneyFlow;
}(ChartIndicator));
export { ChaikinMoneyFlow };
