// @amcharts/amcharts5/.internal/charts/stock/indicators/ChaikinOscillator.js
import { __extends } from "tslib";
import { ChartIndicator } from "./ChartIndicator";
import { LineSeries } from "../../xy/series/LineSeries";
import * as $array from "../../../core/util/Array";
/**
 * An implementation of a [[StockChart]] indicator.
 *
 * @see {@link https://www.amcharts.com/docs/v5/charts/stock/indicators/} for more info
 */
var ChaikinOscillator = /** @class */ (function (_super) {
    __extends(ChaikinOscillator, _super);
    function ChaikinOscillator() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        Object.defineProperty(_this, "_editableSettings", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: [{
                    key: "period",
                    name: _this.root.language.translateAny("Fast period"),
                    type: "number"
                }, {
                    key: "slowPeriod",
                    name: _this.root.language.translateAny("Slow period"),
                    type: "number"
                }, {
                    key: "seriesColor",
                    name: _this.root.language.translateAny("Color"),
                    type: "color"
                }]
        });
        Object.defineProperty(_this, "_themeTag", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "chaikinoscillator"
        });
        return _this;
    }
    Object.defineProperty(ChaikinOscillator.prototype, "_createSeries", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            return this.panel.series.push(LineSeries.new(this._root, {
                xAxis: this.xAxis,
                yAxis: this.yAxis,
                valueXField: "valueX",
                valueYField: "cmf",
                stroke: this.get("seriesColor"),
                fill: undefined
            }));
        }
    });
    Object.defineProperty(ChaikinOscillator.prototype, "_prepareChildren", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            if (this.isDirty("volumeSeries") || this.isDirty("slowPeriod")) {
                this._dataDirty = true;
                this.setCustomData("slowPeriod", this.get("slowPeriod"));
            }
            _super.prototype._prepareChildren.call(this);
        }
    });
    /**
     * @ignore
     */
    Object.defineProperty(ChaikinOscillator.prototype, "prepareData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            if (this.series) {
                var dataItems = this.get("stockSeries").dataItems;
                var volumeSeries_1 = this.get("volumeSeries");
                this.setRaw("field", "close");
                var i_1 = 0;
                var data_1 = this._getDataArray(dataItems);
                var prevAD_1 = 0;
                $array.each(dataItems, function (dataItem) {
                    var close = dataItem.get("valueY");
                    if (close != null) {
                        var low = dataItem.get("lowValueY", close);
                        var high = dataItem.get("highValueY", close);
                        var volume = 1;
                        var volumeDI = volumeSeries_1.dataItems[i_1];
                        if (volumeDI) {
                            volume = volumeDI.get("valueY", 1);
                        }
                        var mf = ((close - low) - (high - close)) / (high - low);
                        if (high == low) {
                            mf = 0;
                        }
                        var ad = prevAD_1 + mf * volume;
                        data_1[i_1].ad = ad;
                        prevAD_1 = ad;
                    }
                    i_1++;
                });
                this._ema(data_1, this.get("slowPeriod", 10), "ad", "emaSlow");
                this._ema(data_1, this.get("period", 3), "ad", "emaFast");
                $array.each(data_1, function (dataItem) {
                    var emaSlow = dataItem.emaSlow;
                    var emaFast = dataItem.emaFast;
                    if (emaSlow != null && emaFast != null) {
                        dataItem.cmf = emaFast - emaSlow;
                    }
                });
                this.series.data.setAll(data_1);
            }
        }
    });
    Object.defineProperty(ChaikinOscillator, "className", {
        enumerable: true,
        configurable: true,
        writable: true,
        value: "ChaikinOscillator"
    });
    Object.defineProperty(ChaikinOscillator, "classNames", {
        enumerable: true,
        configurable: true,
        writable: true,
        value: ChartIndicator.classNames.concat([ChaikinOscillator.className])
    });
    return ChaikinOscillator;
}(ChartIndicator));
export { ChaikinOscillator };
